fix: Add method to get image features in PaliGemmaForConditionalGeneration

YushunXiang opened the pull request:
# What does this PR do?

In the [v4.52.1 release](https://github.com/huggingface/transformers/releases/tag/v4.52.1) of the transformers library, PR #37033 @zucchini-nlp  introduced a bug by renaming the class PaliGemmaForConditionalGeneration(PaliGemmaPreTrainedModel, GenerationMixin) to class PaliGemmaModel(PaliGemmaPreTrainedModel), which causes the original `get_image_features` function in line 218, [huggingface/lerobot/common/policies/pi0/paligemma_with_expert.py](https://github.com/huggingface/lerobot/blob/main/lerobot/common/policies/pi0/paligemma_with_expert.py) to be unusable.

This pull request adds a new `get_image_features` method across multiple generative model implementations in the `src/transformers/models` directory. The method provides a standardized interface for extracting image features from models, with variations in parameters depending on the specific model's requirements.

I modify 6 files with adding the method `get_image_features` to corresponding class `<model name>ForConditionalGeneration`:

- src/transformers/models/idefics2/modeling_idefics2.py
- src/transformers/models/llava/modeling_llava.py
- src/transformers/models/qwen2_vl/modeling_qwen2_vl.py
- src/transformers/models/chameleon/modeling_chameleon.py
- src/transformers/models/paligemma/modeling_paligemma.py
- src/transformers/models/video_llava/modeling_video_llava.py

and use make fix-copies to generate the other 13 modeling files.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts, @qubvel, @zucchini-nlp

YushunXiang committed:
fix: Add method to retrieve image features in PaliGemmaForConditionalGeneration

zucchini-nlp reviewed (commented):
Nice catch! Indeed it is a breaking change and we need to allow access through the generative model as well

Can you update other VLMs as well, since the changes were done on all models?

YushunXiang commented:
> Nice catch! Indeed it is a breaking change and we need to allow access through the generative model as well
> 
> Can you update other VLMs as well, since the changes were done on all models?

Yes, I would like to be a contributor to transformers.

Besides, I have a question is that:

```python
if hasattr(self.paligemma, "get_image_features"):
    return self.paligemma.get_image_features(image)
else:
    return self.paligemma.model.get_image_features(image)
```

With this modification I can run the generated models no matter what `transformers` version, but is it a good modification?

zucchini-nlp commented:
> With this modification I can run the generated models no matter what transformers version, but is it a good modification?

yeah, it can be used as a hacky workaround in your code for `transformers==v4.52`. And for the transformers codebase, your solution looks good and can be propagated to all models. We'll then include your fixes for the next release :)

Btw, after the fixes you need to run `make fix-copies` to make our CI happy

YushunXiang committed:
feat: Add get_image_features method to multiple models for image feature extraction

YushunXiang committed:
Merge branch 'huggingface:main' into fix-paligemma

YushunXiang commented:
I modify 6 files with adding the method `get_image_features` to corresponding class `<model name>ForConditionalGeneration`:

- src/transformers/models/idefics2/modeling_idefics2.py
- src/transformers/models/llava/modeling_llava.py
- src/transformers/models/qwen2_vl/modeling_qwen2_vl.py
- src/transformers/models/chameleon/modeling_chameleon.py
- src/transformers/models/paligemma/modeling_paligemma.py
- src/transformers/models/video_llava/modeling_video_llava.py

and use `make fix-copies` to generate the other 13 `modeling` file.

YushunXiang committed:
fix: reformat the files with ruff.

zucchini-nlp reviewed (commented):
Thanks a lot ❤️ 

Just left a few comments about missing parts, some models use different helpers for video modality or quantized vision tokens

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_38730). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

YushunXiang committed:
feat: Add methods for packing and retrieving image and video features across multiple models

modified:
- modeling_chameleon.py
- modeling_llava_next.py
- modular_llava_next_video.py
- modeling_qwen2_vl.py

and generate the:
- modeling_llava_next_video.py
- modeling_llava_onevision.py
- modeling_qwen2_5_vl.py

YushunXiang commented on src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:
```diff
@@ -1503,6 +1503,14 @@ def set_decoder(self, decoder):
+        vision_feature_layer: Optional[Union[int, List[int]]] = None,
+        vision_feature_select_strategy: Optional[str] = None,
+    ):
+        return self.model.get_image_features(pixel_values_images, vision_feature_layer, vision_feature_select_strategy)
```
Oh, this is my fault. Thanks!

zucchini-nlp reviewed (approved):
Perfect! I think the last rebase went wring, I see unrelated commits in the history. Can you fix that and we'll merge :)

YushunXiang committed:
feat: Implement get_image_features method in Aria, Mistral3, and VipLlava models with updated parameters

YushunXiang commented:
Commit [67461fb](https://github.com/huggingface/transformers/pull/38730/commits/67461fb02aecbeabb13b5e2379fd8228776208e3)
implement `get_image_features` method in `Aria`, `Mistral3`, and `VipLlava` models with updated parameters.

Let me explain the reason for this commit.

In the file `src/transformers/models/aria/modular_aria.py`, `AriaForConditionalGeneration` inherits from `LlavaForConditionalGeneration`. However, the `get_image_features` interface in `AriaModel` is inconsistent with the `get_image_features` interface in `LlavaForConditionalGeneration`.

```python
class AriaModel(LlavaModel):
    def get_image_features(
        self,
        pixel_values: torch.FloatTensor,
        pixel_mask: Optional[torch.FloatTensor] = None,
        vision_feature_layer: int = -1,
    ):
        ...
```


```python
class LlavaModel(LlavaPreTrainedModel):
    def get_image_features(
        self,
        pixel_values: torch.FloatTensor,
        vision_feature_layer: Optional[Union[int, List[int]]] = None,
        vision_feature_select_strategy: Optional[str] = None,
        **kwargs,
    ):
        ...
```


The same issue applies to other files as well.

YushunXiang committed:
fix: reformatted the code with fix-style
